### DISCUSSION TITLE
Apache Pulsar implementation in a distributed system

### DISCUSSION THREAD

**Discussion by SushmitaVelan** [apache/pulsar#21526] [Q&A]
Hi,

I'm new to Apache Pulsar technology and I want to implement the Apache Pulsar in distributed system. In which one machine should have Pulsar Producer which produces the messages to the topic and the other machine should have Pulsar Consumer to consume those messages from the same topic. 

Is there any way to implement the same using C# language. 
Kindly help me with the guideline to implement it!

**Comment by jak78** [comment]
Hi @SushmitaVelan, and welcome to the community!
You can run the producers and consumers on any machine that can connect to the Pulsar cluster. What you should do depends on many factors, including your use case and your expected workload. Could you elaborate?
You can use the [DotPulsar client library](https://pulsar.apache.org/docs/3.1.x/client-libraries-dotnet/) to implement Pulsar producers & consumers in C#.
Hope that helps,

**Reply by SushmitaVelan** [reply]
Hi @jak78, Thank you for your response.

I would like to establish a communication between the 2 pulsar applications across different machine as I've mentioned above.

I've successfully set up the pulsar cluster and started running the pulsar producer in one machine and also, I've started the pulsar consumer in another machine configured the same topic name, same service URL and both the machines are in the same network but still, I couldn't establish the communication.

Is there any configuration in the pulsar configuration file I've to change?

Kindly let me know what I'm missing!
Thank You!

**Reply by jak78** [reply]
Hi! Feel free to share what you observe in the logs so we can help you.

**Reply by SushmitaVelan** [reply]
Hi @jak78,

I've done the below changes. Kindly guide me where am I going wrong !

1)The Producer code (Machine A)

string topic = "pulsar-topic";
            var client = PulsarClient.Builder().ServiceUrl(new Uri("pulsar://Machine A IP:6650")).Build();//pulsar://localhost:6650/
            
            var producer1 = client.NewProducer()
                     .Topic(topic)
                     .Create();

            var producer2 = client.NewProducer()
                     .Topic(topic)
                     .Create();


            for (int i = 0; i < 1000; i++)
            {
                string message = $"Message {i}";
                await producer1.Send(Encoding.UTF8.GetBytes(message + "from Producer1"));
                await producer2.Send(Encoding.UTF8.GetBytes(message + "from Producer2"));
                Console.WriteLine($"Sent: {message}");
            }
            
 The Consumer Coder(Machine B)
 
  string topic = "pulsar-topic";

            var client = PulsarClient.Builder().ServiceUrl(new Uri("pulsar://Machine A IP:6650")).Build();

            var consumer = client.NewConsumer()
                .Topic(topic)
                .SubscriptionType(SubscriptionType.Exclusive)
                .SubscriptionName("shared-subscription1").Create();

            await foreach (var message in consumer.Messages())
            {
                string msg = Encoding.UTF8.GetString(message.Data);
                Console.WriteLine($"Received: {msg}");
               
                await consumer.Acknowledge(message);
            }

2)   I've started the Pulsar cluster in  the both Machines(A & B) and it is running

3)   In Broker.Conf File
      advertisedAddress of Machine A :  <Machine A IP Address>
      advertisedAddress of Machine B :  <Machine B IP Address>

**Reply by jak78** [reply]
Thanks for sharing. I would suggest testing the following:
- Looking for relevant error messages in the consumer and producer logs
- Testing the connection with the Pulsar cluster locally. For example, if a Pulsar broker node runs on MachineA, then open a shell on MachineA and run the following command:
```bash
# Replace PULSAR_HOME with your Pulsar installation path
PULSAR_HOME/bin/pulsar-client produce persistent://public/default/test -m test
```
You should see the following:
```
1 messages successfully produced
```
- Double-checking the `6650` port is open and is reachable from the other machines. You might have a firewall that prevents opening a TCP connection to `6650`.

**Reply by SushmitaVelan** [reply]
Hi @jak78 , 
Thank you for your response! 
I'll try the steps which you have mentioned..

**Comment by SushmitaVelan** [comment]
Hi @jak78 , 
From your suggestion I went through the logs and I found out when I try to deploy the Zookeeper cluster with this command "bin/pulsar-daemon start zookeeper" in WSL Ubuntu machine. 

The below steps I've done for deploying Zookeeper cluster in each of my 3 machine's WSL Ubuntu(22.04):

1) In conf/zookeeper.conf, added the ipaddress of those 3 window machine's
# ZooKeeper servers
server.1=192.168.135.190:2888:3888
server.2=192.168.135.202:2888:3888
server.3=192.168.135.101:2888:3888

2)mkdir -p /data/zookeeper
echo 1 | sudo tee /demo/zookeeper/myid

3) bin/pulsar-daemon start zookeeper

I got the following errors:
<img width="935" alt="error" src="https://github.com/apache/pulsar/assets/119869052/8a58e2e2-97be-488a-87ce-12c48ffb0e28">

Kindly guide me regarding this
Thanks!

**Reply by lhotari** [reply]
Deploying Pulsar in production on WSL isn't a supported configuration. For typical development usage, you can simply run Pulsar Standalone in Docker with this command:
```
docker run -it -p 8080:8080 -p 6650:6650 apachepulsar/pulsar /pulsar/bin/pulsar standalone -nss -nfw
```

**Reply by SushmitaVelan** [reply]
Thank you for your reply @lhotari !
I want to deploy the Apache Pulsar on bare metal.
I did the following steps: 
1) I've installed Ubuntu VM in three machines
2)I've started the Zookeeper cluster in all those 3 machines and status is started
3) In the initialize-cluster-metadata steps, I'm facing the below error
This is the Command:
bin/pulsar initialize-cluster-metadata \
    --cluster pulsar-cluster-1 \
    --metadata-store zk:10.0.2.151:2181,10.0.2.152:2181,10.0.2.153:2181 \
    --configuration-metadata-store zk:10.0.2.151:2181,10.0.2.152:2181,10.0.2.153:2181 \
    --web-service-url http://10.0.2.151:8080,10.0.2.152:8080,10.0.2.153:8080 \
    --web-service-url-tls https://10.0.2.151:8443,10.0.2.152:8443,10.0.2.153:8443 \
    --broker-service-url pulsar://10.0.2.151:6650,10.0.2.152:6650,10.0.2.153:6650 \
    --broker-service-url-tls pulsar+ssl://10.0.2.151:6651,10.0.2.152:6651,10.0.2.153:6651
    
    The error I'm facing is:
    
 ![1](https://github.com/apache/pulsar/assets/119869052/034379b0-6038-40ae-9cfa-4b319036f3f4)
![2](https://github.com/apache/pulsar/assets/119869052/7e6d6aeb-3614-4a1c-84d9-059ddb787ebd)

Kindly guide me what is wrong?

**Reply by jak78** [reply]
I can see a `no route to host` error in the logs.
This indicates a network configuration issue, unrelated to ZooKeeper.
Can you make sure the three VMs can connect to each other over the network? What happens when you run a `ping` command from one VM to another?

**Reply by SushmitaVelan** [reply]
Hi Jak78,
Thank you so much for your reply.
I've successfully implemented it, It was purely network configuration issue only.

**Reply by jak78** [reply]
@SushmitaVelan I'm delighted that you resolved the problem. Have a wonderful journey through Pulsar!
